python/variable.cpp

```diff
@@ -23,34 +23,52 @@
 #include "bind_data_access.h"
 #include "bind_operators.h"
 #include "bind_slice_methods.h"
-#include "docstring.h"
 #include "dtype.h"
 #include "make_variable.h"
 #include "numpy.h"
+#include "py_object.h"
 #include "pybind11.h"
 #include "rename.h"
-#include "unit.h"
 
 using namespace scipp;
 using namespace scipp::variable;
 
 namespace py = pybind11;
 
+namespace {
+template <class T> void check_dtype_matches(const py::object &dtype) {
+  const auto requested_dtype = scipp_dtype(dtype);
+  const auto data_dtype =
+      scipp::dtype<std::conditional_t<std::is_same_v<T, py::object>,
+                                      scipp::python::PyObject, T>>;
+  if (requested_dtype != data_dtype) {
+    throw std::invalid_argument("Cannot convert values of type " +
+                                to_string(data_dtype) + " to " +
+                                to_string(requested_dtype));
+  }
+}
+} // namespace
+
 template <class T> void bind_init_0D(py::class_<Variable> &c) {
+  // Conversions are in general not possible. We have the dtype
+  // arguments only for symmetry with other constructors.
```

Comment:
Can you elaborate? Can't we convert, say, `int` to `double`?

Reply:
Yes. But they are handled by `bind_init_0D_native_python_types`. This function is for cases where we cannot convert (so easily / reasonable).